File: src/main/java/com/globalcarelink/review/Review.java
package com.globalcarelink.review;

import com.globalcarelink.auth.Member;
import com.globalcarelink.common.entity.BaseEntity;
import com.globalcarelink.facility.FacilityProfile;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * 리뷰 엔티티
 * 시설 및 서비스에 대한 사용자 리뷰 관리
 */
@Entity
@Table(name = "reviews")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Review extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * 리뷰 작성자
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reviewer_id", nullable = false)
    private Member reviewer;

    /**
     * 리뷰 대상 시설
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "facility_id", nullable = false)
    private FacilityProfile facility;

    /**
     * 리뷰 제목
     */
    @Column(nullable = false, length = 200)
    private String title;

    /**
     * 리뷰 내용
     */
    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    /**
     * 전체 평점 (1.0 ~ 5.0)
     */
    @Column(nullable = false, precision = 2, scale = 1)
    private BigDecimal overallRating;

    /**
     * 서비스 품질 평점
     */
    @Column(precision = 2, scale = 1)
    private BigDecimal serviceQualityRating;

    /**
     * 시설 환경 평점
     */
    @Column(precision = 2, scale = 1)
    private BigDecimal facilityRating;

    /**
     * 직원 친절도 평점
     */
    @Column(precision = 2, scale = 1)
    private BigDecimal staffRating;

    /**
     * 가격 만족도 평점
     */
    @Column(precision = 2, scale = 1)
    private BigDecimal priceRating;

    /**
     * 접근성 평점
     */
    @Column(precision = 2, scale = 1)
    private BigDecimal accessibilityRating;

    /**
     * 리뷰 타입
     */
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ReviewType reviewType = ReviewType.FACILITY;

    /**
     * 리뷰 상태
     */
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ReviewStatus status = ReviewStatus.ACTIVE;

    /**
     * 추천 여부 (시설을 다른 사람에게 추천하겠는가)
     */
    @Column(nullable = false)
    private Boolean recommended = true;

    /**
     * 방문 날짜
     */
    @Column
    private LocalDateTime visitDate;

    /**
     * 서비스 이용 기간 (일 단위)
     */
    @Column
    private Integer serviceDurationDays;

    /**
     * 도움이 된다고 투표한 수
     */
    @Column(nullable = false)
    private Integer helpfulCount = 0;

    /**
     * 도움이 안 된다고 투표한 수
     */
    @Column(nullable = false)
    private Integer notHelpfulCount = 0;

    /**
     * 신고 횟수
     */
    @Column(nullable = false)
    private Integer reportCount = 0;

    /**
     * 관리자 응답
     */
    @Column(columnDefinition = "TEXT")
    private String adminResponse;

    /**
     * 관리자 응답 일시
     */
    @Column
    private LocalDateTime adminResponseDate;

    /**
     * 관리자 응답 작성자
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "admin_responder_id")
    private Member adminResponder;

    /**
     * 검증된 리뷰 여부 (실제 이용자 확인)
     */
    @Column(nullable = false)
    private Boolean verified = false;

    /**
     * 익명 리뷰 여부
     */
    @Column(nullable = false)
    private Boolean anonymous = false;

    /**
     * 첨부 이미지 URL 목록
     */
    @ElementCollection
    @CollectionTable(name = "review_images", joinColumns = @JoinColumn(name = "review_id"))
    @Column(name = "image_url", length = 500)
    private List<String> imageUrls = new ArrayList<>();

    /**
     * 태그 목록 (장점/단점 태그)
     */
    @ElementCollection
    @CollectionTable(name = "review_tags", joinColumns = @JoinColumn(name = "review_id"))
    @Column(name = "tag", length = 50)
    private List<String> tags = new ArrayList<>();

    /**
     * 리뷰 타입 열거형
     */
    public enum ReviewType {
        FACILITY("시설 리뷰"),
        SERVICE("서비스 리뷰"),
        CAREGIVER("요양보호사 리뷰"),
        PROGRAM("프로그램 리뷰");

        private final String displayName;

        ReviewType(String displayName) {
            this.displayName = displayName;
        }

        public String getDisplayName() {
            return displayName;
        }
    }

    /**
     * 리뷰 상태 열거형
     */
    public enum ReviewStatus {
        ACTIVE("활성"),
        PENDING("검토중"),
        BLOCKED("차단됨"),
        DELETED("삭제됨");

        private final String displayName;

        ReviewStatus(String displayName) {
            this.displayName = displayName;
        }

        public String getDisplayName() {
            return displayName;
        }
    }

    /**
     * 리뷰어 이름 반환 (익명 처리 고려)
     */
    public String getReviewerDisplayName() {
        if (this.anonymous || this.reviewer == null) {
            return "익명";
        }
        return this.reviewer.getName();
    }

    /**
     * 시설명 반환 (안전한 접근)
     */
    public String getFacilityName() {
        return this.facility != null ? this.facility.getName() : "알 수 없음";
    }

    /**
     * 평균 세부 평점 계산
     */
    public BigDecimal getAverageDetailRating() {
        int count = 0;
        BigDecimal sum = BigDecimal.ZERO;

        if (serviceQualityRating != null) {
            sum = sum.add(serviceQualityRating);
            count++;
        }
        if (facilityRating != null) {
            sum = sum.add(facilityRating);
            count++;
        }
        if (staffRating != null) {
            sum = sum.add(staffRating);
            count++;
        }
        if (priceRating != null) {
            sum = sum.add(priceRating);
            count++;
        }
        if (accessibilityRating != null) {
            sum = sum.add(accessibilityRating);
            count++;
        }

        if (count == 0) {
            return BigDecimal.ZERO;
        }

        return sum.divide(BigDecimal.valueOf(count), 1, BigDecimal.ROUND_HALF_UP);
    }


    /**
     * 이미지 첨부 여부
     */
    public boolean hasImages() {
        return imageUrls != null && !imageUrls.isEmpty();
    }

    /**
     * 태그 존재 여부
     */
    public boolean hasTags() {
        return tags != null && !tags.isEmpty();
    }

    /**
     * 관리자 응답 존재 여부
     */
    public boolean hasAdminResponse() {
        return adminResponse != null && !adminResponse.trim().isEmpty();
    }

    /**
     * 도움됨 투표 증가
     */
    public void incrementHelpfulVote() {
        this.helpfulCount = (this.helpfulCount != null ? this.helpfulCount : 0) + 1;
    }

    /**
     * 도움됨 투표 감소
     */
    public void decrementHelpfulVote() {
        this.helpfulCount = Math.max(0, (this.helpfulCount != null ? this.helpfulCount : 0) - 1);
    }

    /**
     * 도움 안됨 투표 증가
     */
    public void incrementNotHelpfulVote() {
        this.notHelpfulCount = (this.notHelpfulCount != null ? this.notHelpfulCount : 0) + 1;
    }

    /**
     * 도움 안됨 투표 감소
     */
    public void decrementNotHelpfulVote() {
        this.notHelpfulCount = Math.max(0, (this.notHelpfulCount != null ? this.notHelpfulCount : 0) - 1);
    }

    /**
     * 신고 횟수 증가
     */
    public void incrementReportCount() {
        this.reportCount = (this.reportCount != null ? this.reportCount : 0) + 1;
        
        // 신고 횟수가 임계값을 초과하면 자동으로 검토 상태로 변경
        if (this.reportCount >= 5) {
            this.status = ReviewStatus.PENDING;
        }
    }

    /**
     * 관리자 응답 추가
     */
    public void addAdminResponse(String response, Member admin) {
        this.adminResponse = response;
        this.adminResponseDate = LocalDateTime.now();
        this.adminResponder = admin;
    }

    /**
     * 리뷰 차단
     */
    public void block() {
        this.status = ReviewStatus.BLOCKED;
    }

    /**
     * 리뷰 활성화
     */
    public void activate() {
        this.status = ReviewStatus.ACTIVE;
    }

    /**
     * 리뷰 삭제 (soft delete)
     */
    public void delete() {
        this.status = ReviewStatus.DELETED;
    }

    /**
     * 검증 상태 설정
     */
    public void setVerified(boolean verified) {
        this.verified = verified;
    }

    /**
     * 이미지 추가
     */
    public void addImageUrl(String imageUrl) {
        if (this.imageUrls == null) {
            this.imageUrls = new ArrayList<>();
        }
        this.imageUrls.add(imageUrl);
    }

    /**
     * 태그 추가
     */
    public void addTag(String tag) {
        if (this.tags == null) {
            this.tags = new ArrayList<>();
        }
        if (!this.tags.contains(tag)) {
            this.tags.add(tag);
        }
    }

    /**
     * 활성 리뷰 여부 확인
     */
    public boolean isActive() {
        return this.status == ReviewStatus.ACTIVE;
    }

    /**
     * 수정 가능 여부 확인 (작성자 본인만 가능, 24시간 이내)
     */
    public boolean isEditable(Member member) {
        if (!this.reviewer.equals(member)) {
            return false;
        }
        
        if (this.status != ReviewStatus.ACTIVE) {
            return false;
        }
        
        // 24시간 이내만 수정 가능
        LocalDateTime cutoff = this.getCreatedDate().plusHours(24);
        return LocalDateTime.now().isBefore(cutoff);
    }
}

File: frontend-backup-20250821_151959/src/types/review.ts
/**
 * 리뷰 시스템 타입 정의
 * 백엔드 Review 엔티티와 완전히 동기화
 */

/**
 * 리뷰 타입 열거형
 */
export enum ReviewType {
  FACILITY = 'FACILITY',        // 시설 리뷰
  SERVICE = 'SERVICE',          // 서비스 리뷰
  CAREGIVER = 'CAREGIVER',      // 요양보호사 리뷰
  PROGRAM = 'PROGRAM'           // 프로그램 리뷰
}

/**
 * 리뷰 상태 열거형
 */
export enum ReviewStatus {
  ACTIVE = 'ACTIVE',            // 활성
  PENDING = 'PENDING',          // 검토중
  BLOCKED = 'BLOCKED',          // 차단됨
  DELETED = 'DELETED'           // 삭제됨
}

/**
 * 리뷰 투표 타입
 */
export enum VoteType {
  HELPFUL = 'HELPFUL',          // 도움됨
  NOT_HELPFUL = 'NOT_HELPFUL'   // 도움안됨
}

/**
 * 신고 사유
 */
export enum ReportReason {
  INAPPROPRIATE_CONTENT = 'INAPPROPRIATE_CONTENT',  // 부적절한 내용
  SPAM = 'SPAM',                                    // 스팸
  FALSE_INFORMATION = 'FALSE_INFORMATION',          // 거짓 정보
  OFFENSIVE_LANGUAGE = 'OFFENSIVE_LANGUAGE',        // 욕설/비방
  PRIVACY_VIOLATION = 'PRIVACY_VIOLATION',          // 개인정보 침해
  COPYRIGHT_VIOLATION = 'COPYRIGHT_VIOLATION',      // 저작권 침해
  OTHER = 'OTHER'                                   // 기타
}

/**
 * 리뷰 정렬 옵션
 */
export enum ReviewSort {
  LATEST = 'latest',            // 최신순
  OLDEST = 'oldest',            // 오래된순
  HIGHEST_RATING = 'highest',   // 평점 높은순
  LOWEST_RATING = 'lowest',     // 평점 낮은순
  MOST_HELPFUL = 'helpful'      // 도움됨 많은순
}

/**
 * 리뷰 필터 옵션
 */
export interface ReviewFilter {
  reviewType?: ReviewType;      // 리뷰 타입
  minRating?: number;           // 최소 평점
  maxRating?: number;           // 최대 평점
  verified?: boolean;           // 검증된 리뷰만
  hasImages?: boolean;          // 이미지 포함 리뷰만
  recommended?: boolean;        // 추천 리뷰만
}

/**
 * 기본 리뷰 인터페이스
 */
export interface Review {
  id: number;
  reviewer: {
    id: number;
    name: string;
    email?: string;
  };
  facility: {
    id: number;
    name: string;
    type: string;
  };
  title: string;
  content: string;
  overallRating: number;
  serviceQualityRating?: number;
  facilityRating?: number;
  staffRating?: number;
  priceRating?: number;
  accessibilityRating?: number;
  reviewType: ReviewType;
  status: ReviewStatus;
  recommended: boolean;
  visitDate?: string;
  serviceDurationDays?: number;
  helpfulCount: number;
  notHelpfulCount: number;
  reportCount: number;
  adminResponse?: string;
  adminResponseDate?: string;
  adminResponder?: {
    id: number;
    name: string;
  };
  verified: boolean;
  anonymous: boolean;
  imageUrls: string[];
  tags: string[];
  createdDate: string;
  modifiedDate: string;
  
  // 계산된 속성
  reviewerDisplayName?: string;
  averageDetailRating?: number;
  hasImages?: boolean;
  hasTags?: boolean;
  hasAdminResponse?: boolean;
  isEditable?: boolean;
}

/**
 * 리뷰 작성 요청
 */
export interface ReviewCreateRequest {
  title: string;
  content: string;
  overallRating: number;
  serviceQualityRating?: number;
  facilityRating?: number;
  staffRating?: number;
  priceRating?: number;
  accessibilityRating?: number;
  reviewType: ReviewType;
  recommended: boolean;
  visitDate?: string;
  serviceDurationDays?: number;
  anonymous: boolean;
  imageUrls?: string[];
  tags?: string[];
}

/**
 * 리뷰 수정 요청
 */
export interface ReviewUpdateRequest {
  title?: string;
  content?: string;
  overallRating?: number;
  serviceQualityRating?: number;
  facilityRating?: number;
  staffRating?: number;
  priceRating?: number;
  accessibilityRating?: number;
  recommended?: boolean;
  anonymous?: boolean;
  imageUrls?: string[];
  tags?: string[];
}

/**
 * 리뷰 투표 인터페이스
 */
export interface ReviewVote {
  id: number;
  reviewId: number;
  voterId: number;
  voteType: VoteType;
  createdDate: string;
}

/**
 * 리뷰 신고 인터페이스
 */
export interface ReviewReport {
  id: number;
  reviewId: number;
  reporterId: number;
  reason: ReportReason;
  description?: string;
  status: 'PENDING' | 'RESOLVED' | 'REJECTED';
  createdDate: string;
  resolvedDate?: string;
}

/**
 * 리뷰 통계
 */
export interface ReviewStatistics {
  totalCount: number;
  averageRating: number;
  recommendationPercentage: number;
  ratingDistribution: {
    5: number;
    4: number;
    3: number;
    2: number;
    1: number;
  };
  categoryRatings: {
    serviceQuality: number;
    facility: number;
    staff: number;
    price: number;
    accessibility: number;
  };
}

/**
 * 리뷰 응답 (API 응답)
 */
export interface ReviewResponse {
  content: Review[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
  empty: boolean;
}

/**
 * 관리자 응답 요청
 */
export interface AdminResponseRequest {
  response: string;
}

/**
 * 리뷰 태그 옵션
 */
export const REVIEW_TAGS = {
  positive: [
    '친절한 직원',
    '깨끗한 시설',
    '좋은 위치',
    '합리적인 가격',
    '우수한 서비스',
    '편리한 접근성',
    '다양한 프로그램',
    '전문적인 케어',
    '맛있는 식사',
    '안전한 환경'
  ],
  negative: [
    '불친절한 직원',
    '시설 노후',
    '접근성 불편',
    '비싼 가격',
    '서비스 부족',
    '위생 문제',
    '소통 부족',
    '프로그램 부족',
    '식사 불만족',
    '안전 우려'
  ]
};

/**
 * 평점 단계별 표시
 */
export const RATING_LABELS: Record<number, string> = {
  5: '매우 만족',
  4: '만족',
  3: '보통',
  2: '불만족',
  1: '매우 불만족'
};

/**
 * 리뷰 타입별 표시명
 */
export const REVIEW_TYPE_LABELS: Record<ReviewType, string> = {
  [ReviewType.FACILITY]: '시설 리뷰',
  [ReviewType.SERVICE]: '서비스 리뷰',
  [ReviewType.CAREGIVER]: '요양보호사 리뷰',
  [ReviewType.PROGRAM]: '프로그램 리뷰'
};

/**
 * 신고 사유별 표시명
 */
export const REPORT_REASON_LABELS: Record<ReportReason, string> = {
  [ReportReason.INAPPROPRIATE_CONTENT]: '부적절한 내용',
  [ReportReason.SPAM]: '스팸',
  [ReportReason.FALSE_INFORMATION]: '거짓 정보',
  [ReportReason.OFFENSIVE_LANGUAGE]: '욕설/비방',
  [ReportReason.PRIVACY_VIOLATION]: '개인정보 침해',
  [ReportReason.COPYRIGHT_VIOLATION]: '저작권 침해',
  [ReportReason.OTHER]: '기타'
};